### PR TITLE
Added real time reporting to spec reporter 

### LINKED ISSUE #5056: Reporter prints output only after all suites are finished
**Describe your question with as much detail as possible**
After migration from webdriver 4 to 5 - reporter output is shown only after all suites are finished running. Before, it used to print "as it goes". How can I configure wdio to print reporter output to console as soon as single test is finished?

**Environment (please complete the following information):**
 - **WebdriverIO version:** 5.18.7
 - **Mode:** Wdio test runner (local)
 - **If WDIO Testrunner, running sync/async:** sync
 - **Node.js version:** 10.19.0
 - **NPM version:** 6.13.4
 - **Browser name and version:** Chrome 80
 - **Platform name and version:** Windows 10
 - **Additional wdio packages used (if applicable):** @wdio/spec reporter, wdio-teamcity-reporter

**Additional context**

I've looked into the code and found that CliInterface would store messages from reporters unless you run wdio in --watch mode. Does it mean there's a need for feature request?

### LINKED ISSUE #7532: Allow real time reporting of reporters using WDIO Testrunner
**DD** 
this may be a duplicate(apologies if it is) i  did a search in all issues, found 25 closed issues that may relate to spec reporter in real time but couldnt find much information except for a comment that stated watchmode, i posted in gitter but no response as of yet so i just want to see if someone out there has found something i have yet to find regarding real time output of the spec being ran, I also searched all of gitter for solutions but came up short, i looked through the webdriverio documentation but couldnt find much info about real time reporters

**Is your feature request related to a problem? Please describe.**
 I'm always frustrated when i have to wait until the end of the run to see output

**Describe the solution you'd like**
im coming over from protractor, and theres a npm package for real time reporters, id like to see something similar

**Describe alternatives you've considered**
I have done global searches inside gitter and saw some people put code in the afterStep but from my understanding that is used for cucumber or the framework im not using(which is jasmine). please keep me honest here!

### PR DESCRIPTION
**Pull request by praveendvd**
## Proposed changes

This pull request logs the current test status (Pass , failed ,skipped)  for spec reporter .

This will fix : 

Allow real time reporting of reporters using WDIO Testrunner #7532

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

Test will be added 

Fixes https://github.com/webdriverio/webdriverio/issues/7532
Fixes #5056

### CODE REVIEW

**Comment by praveendvd** [comment]
https://github.com/webdriverio/webdriverio/issues/7532#issuecomment-947176949

@christian-bromann

**Comment by praveendvd** [comment]
> @praveendvd @christian-bromann are you sure that for `maxInstances> 1...100` the terminal logs won't become a terrible mess?
> 
> It is riduculus that we call console.log() when we have @wdio/logger...

Could you let me know how to use the logger ? i used this.write but it flushes the output only after the execution completes. Doesnt behave in a realtime manner

**Comment by BorisOsipov** [comment]
>Could you let me know how to use the logger ?

I don't know. Need investigation why it doesn't print "realtime". I think if logger is not worked as you desired it is not the reason to use console.log().

**Comment by praveendvd** [comment]
@BorisOsipov  @christian-bromann  replaced console.log with logger and explicit log.level info

**Comment by christian-bromann** [comment]
@praveendvd sorry for not reviewing this earlier. Feel free to always ping me directly in Gitter or in the issue thread.

I think we can move forward with this with 2 additional changes:
- rebase on top of current `main` branch
- replace logger and enhance worker output logic - if you use logger and the user has `outputDir` set, then everything is being logged into a file rather than into stdout. Normally all reporter use `this.write`. `write` is implemented in the base reporter but prints everything into a write stream. I suggest to enhance the [CLI interface](https://github.com/webdriverio/webdriverio/blob/f26594ce69498bdf7365485347b1487f7f80f873/packages/wdio-cli/src/interface.ts) and allow reporters to print to stdout immediately - I can see that this is desired for other reporters too, e.g. dot reporter

**Comment by praveendvd** [comment]
> * implemented

HI @christian-bromann  as mentioned this.write flushes everything to stdout only at the end of execution , thats why used logger,  I am not aware what changes to make in cli interface could you shed some more information

**Comment by praveendvd** [comment]
@christian-bromann  COuld you have a look at the Mac error , its there for both of my pull request not sure why

**Comment by christian-bromann** [comment]
> thats why used logger, I am not aware what changes to make in cli interface could you shed some more information

As mentioned above our logger only works if you don't have `outputDir` defined. Please have a look at [`/packages/wdio-cli/src/interface.ts`](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-cli/src/interface.ts#L209-L252) which allows you to print messages to stdout from the parent process. From the base reporter you should be able to post messages to that parent process and have it handle the message there. I haven't touched the code in a while so I can't point you to concrete code examples.

**Comment by praveendvd** [comment]
@christian-bromann  Changed logger to event , now we send a realtimereport event that will log to console . We can throw this event from any reporter

**Comment by praveendvd** [comment]
I added the change but i strogly believe logger will be much more useful as @BorisOsipov  mentioned. This implementation works well if the loglevel is silent else the status will get hidden in the sea of logs that we get from wdio. IF we use logger , then user can opt of point it to a file if required so that they can see it in a much more cleaner way . 

so the new implementation works exactly the same as if we use logger, but doesnt have option to write to a file. should we implement a flag that allows to write to a file ?

File: packages/wdio-spec-reporter/src/index.ts [hunk 1]
```diff
@@ -106,6 +120,40 @@ export default class SpecReporter extends WDIOReporter {
         this.printReport(runner)
     }
 
+    /**
+     * Print the report to the stdout realtime
+     */
+
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/src/index.ts:126]:
```suggestion
```

File: packages/wdio-spec-reporter/src/index.ts [hunk 2]
```diff
@@ -106,6 +120,40 @@ export default class SpecReporter extends WDIOReporter {
...
+    printCurrentStats (stat: TestStats | HookStats | SuiteStats) {
+        if (!this._realTimeReporting)
+            return
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/src/index.ts:129]:
```suggestion
        if (!this._realTimeReporting) {
            return
        }
```

File: packages/wdio-spec-reporter/src/index.ts [hunk 3]
```diff
@@ -48,6 +51,7 @@ export default class SpecReporter extends WDIOReporter {
 
         this._symbols = { ...this._symbols, ...this.options.symbols || {} }
         this._onlyFailures = options.onlyFailures || false
+        this._realTimeReporting = options.realTimeReporting || false
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/src/index.ts:54]:
```suggestion
        this._realtimeReporting = options.realtimeReporting || false
```

File: packages/wdio-spec-reporter/src/index.ts [hunk 4]
```diff
@@ -18,11 +18,14 @@ export default class SpecReporter extends WDIOReporter {
...
     private _orderedSuites: SuiteStats[] = []
     private _consoleOutput = ''
+    private _suiteIndent = ''
+    private _preface = ''
     private _consoleLogs: string[] = []
     private _originalStdoutWrite = process.stdout.write.bind(process.stdout)
 
     private _addConsoleLogs = false
-
+    private _realTimeReporting = false
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/src/index.ts:27]:
```suggestion
    private _realtimeReporting = false
```

File: packages/wdio-spec-reporter/README.md [hunk 5]
```diff
@@ -98,3 +98,18 @@ Default: `false`
   },
 ]
 ```
+
+### realTimeReporting
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/README.md:102]:
```suggestion
### realtimeReporting
```

File: packages/wdio-spec-reporter/README.md [hunk 6]
```diff
@@ -98,3 +98,18 @@ Default: `false`
...
+Set to true to display test status realtime than just at the end of the run
+
+Type: `boolean`
+Default: `false`
+
+```js
+[
+  "spec",
+  {
+    realTimeReporting: true,
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/README.md:112]:
```suggestion
    realtimeReporting: true,
```

File: packages/wdio-spec-reporter/src/index.ts [hunk 7]
```diff
@@ -106,6 +120,40 @@ export default class SpecReporter extends WDIOReporter {
         this.printReport(runner)
     }
 
+    /**
+     * Print the report to the stdout realtime
+     */
+
+    printCurrentStats (stat: TestStats | HookStats | SuiteStats) {
+        if (!this._realTimeReporting)
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/src/index.ts:128]:
```suggestion
        if (!this._realtimeReporting)
```

File: packages/wdio-spec-reporter/src/types.ts [hunk 8]
```diff
@@ -42,6 +42,12 @@ export interface SpecReporterOptions {
      * @default: false
      */
     addConsoleLogs?: boolean
+    /**
+     * Ability to show test status realtime
+     *
+     * @default: false
+     */
+     realTimeReporting?: boolean
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/src/types.ts:50]:
```suggestion
     realtimeReporting?: boolean
```

File: packages/wdio-spec-reporter/src/types.ts [hunk 9]
```diff
@@ -42,6 +42,12 @@ export interface SpecReporterOptions {
      * @default: false
      */
     addConsoleLogs?: boolean
+    /**
+     * Ability to show test status realtime
+     *
+     * @default: false
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/src/types.ts:48]:
```suggestion
     * @default: `false`
```

File: packages/wdio-spec-reporter/tests/index.test.ts [hunk 10]
```diff
@@ -831,4 +849,84 @@ describe('SpecReporter', () => {
             }, false)).toBe('chrome 50 (unknown)')
         })
     })
+
+    describe('add real time report', () => {
+        const options = { realTimeReporting: true }
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/tests/index.test.ts:854]:
```suggestion
        const options = { realtimeReporting: true }
```

File: packages/wdio-spec-reporter/tests/index.test.ts [hunk 11]
```diff
@@ -831,4 +849,84 @@ describe('SpecReporter', () => {
...
+                title:'test1',
+                state:'skipped'
+            })
+            tmpReporter.onSuiteEnd()
+            tmpReporter.onRunnerEnd(runnerEnd())
+        })
+    })
+
+    it('should call printCurrentStats on Hook complete', () => {
+        tmpReporter = new SpecReporter({ realTimeReporting : false })
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/tests/index.test.ts:915]:
```suggestion
        tmpReporter = new SpecReporter({ realtimeReporting : false })
```

**Comment by christian-bromann** [comment]
Just added some minor formatting adjustments and renamed the new option from `realTimeReporting` to `realtimeReporting` given that realtime is one word.

File: packages/wdio-spec-reporter/src/index.ts [hunk 12]
```diff
@@ -106,6 +120,40 @@ export default class SpecReporter extends WDIOReporter {
         this.printReport(runner)
     }
 
+    /**
+     * Print the report to the stdout realtime
+     */
+    printCurrentStats (stat: TestStats | HookStats | SuiteStats) {
+        if (!this._realTimeReporting) {
```
Comment (christian-bromann) [inline packages/wdio-spec-reporter/src/index.ts:127]:
```suggestion
        if (!this._realtimeReporting) {
```

**Comment by christian-bromann** [comment]
There seems to be a typing issue with Jest. I will go ahead and merge and check in `main` branch.

**Review by christian-bromann — APPROVED** [review]
Thanks a lot, this is awesome 👍

**Pull request merged**